### DISCUSSION TITLE
Misc notes

### DISCUSSION THREAD

**Discussion by thorwhalen** [addaix/raglab#1] [Ideas]
For miscellaneous notes.

**Comment by thorwhalen** [comment]
## A raglab dag

![image](https://github.com/addaix/raglab/assets/1906276/62ed100b-962b-48f3-8223-528bd117bd8d)

```python
from meshed import code_to_dag

@code_to_dag
def rag_dashboard():
    dataset_list = get_dataset_list(user, pwd)
    dataset = select_dataset(dataset_list, dataset_choice)
    dataset_query = get_dataset_query(query)
    relevant_chunk_keys = get_relevant_chunks(dataset_query, dataset)
    result_display = make_chunks_display(relevant_chunk_keys)
    relevant_chunk_contents = get_content(relevant_chunk_keys)
    relevant_chunk_contents_aggregation = aggregate_content(relevant_chunk_contents)
    answer = make_answer(query, relevant_chunk_contents_aggregation)

rag_dashboard.dot_digraph()
```

**Comment by thorwhalen** [comment]
## VSCode designs I'd like to emulate

Visual Studio Code (VSCode) is a popular text editor and integrated development environment (IDE) that was designed with usability, flexibility, and extensibility in mind. Here's how these principles are reflected in its configuration CRUD (Create, Read, Update, Delete) and command setup:

### Configuration CRUD

In VSCode, user settings are handled in a unique and flexible manner. There are two ways to view and edit configurations: through a GUI and through a JSON file.

* **GUI**: The Settings GUI provides an easy and intuitive interface for users to customize their VSCode experience. Users can search for a specific setting, see its current value, modify it, and even see a description of what it does. This view is friendly to users who prefer graphical interfaces and are not comfortable with direct JSON editing.

* **JSON file**: For users who prefer a code-based approach, VSCode also provides a settings.json file. This JSON file provides a more granular level of control, allowing users to adjust even the most intricate settings. It's also beneficial for power users and developers who may prefer to work directly with code and want to version control their settings.

**--> This is what I'll often refer to as a [store explorer](https://github.com/addaix/raglab/discussions/1#discussioncomment-8517075) pattern (that uses `dol`+`py2http` in the backend)**

### Command Palette

The Command Palette in VSCode (triggered using Ctrl+Shift+P or Cmd+Shift+P on macOS) is a powerful tool that exposes most of the editor's functionality. It provides a quick way to access commands without having to navigate through menus or remember specific keybindings.

* **Fast and Efficient**: The Command Palette employs a fuzzy search mechanism, meaning you don't need to know the exact name of the command. As you type, it suggests commands that match your input, making it highly efficient.
* **Versatility**: The Command Palette can perform a multitude of tasks: from basic operations like opening and closing files, to more complex ones like sorting lines of text or changing the syntax highlighting of the current file.
* **Extensibility**: VsCode is highly extensible. Many extensions can add their commands to the Command Palette, making it a one-stop-shop for most of the actions you'd like to perform in VsCode.

In designing your system, emulating these VSCode features would provide users with a flexible, user-friendly, and efficient way to navigate and customize their experience. Considering a variety of user preferences, like VSCode does with its GUI and JSON settings options, can increase the accessibility and usability of your system. Likewise, a command setup similar to VSCode's Command Palette can streamline user interaction and improve efficiency.

### Packages to use?

There's probably several JS frameworks out there to make the above easy to do. 

A little comparative shopping is in order here. 

I favor using [rjsf](https://rjsf-team.github.io/react-jsonschema-form/) where ever we can though.

**Comment by thorwhalen** [comment]
# Store explorers

Easily "dispatch" a store to a GUI. 

**Note: Follow the more general discussion here: [Collections Operations UI Design Patterns](https://github.com/otosense/oui/discussions/14)**

Many user interactions with our system will be, or can be expressed as, CRUD on a store (a (possibly mutable) `Mapping`). 
We will focus our discussion on this case, but the general context of [Collections Operations UI Design Patterns](https://github.com/otosense/oui/discussions/14) should be understood as well as the general context of object mapping, or more precisely, interface mapping (`i2i`), which is what most of [i2mint](https://github.com/i2mint) is about. 

[dol](https://github.com/i2mint/dol) allows us to normalize most data sources (storage systems etc.) to the `Mapping` interface. 

This means, with [py2http](https://github.com/i2mint/py2http), we can have normalized Web services to interact with our data. Building on that, we should be able to have several UI components that provide data views and explorers. 

Here are a few concrete forms this could take:
* For tabular, or any "fixed schema" data: 
  * A smart-table perspective to meta-data (or any data that has a fixed schema)
  * Cards collections: Example: [Google search results](https://www.google.com/search?q=reusable+ui+components), [image results](https://www.google.com/search?q=reusable+UI+components&source=lnms&tbm=isch&sa=X&biw=1648&bih=920&dpr=2), [shopping results](https://www.google.com/search?q=reusable+UI+components&source=lmns&tbm=shop&bih=920&biw=1648&hl=en), etc.
* For nested data (especially if not fixed schema):
  * A file-explorer-like perspective to blob data. (Note: [stack app](https://github.com/otosense/stack), which was used as the user-facing store explorer to upload and view files.). 
  * configuration CRUD (user preferences, etc.): Think of [the way VSCode does configuration CRUD](https://github.com/addaix/raglab/discussions/1#discussioncomment-8517049)

* "Smart-dynamic-filter". Consider VSCode's command palette, where commands can be found and launched via a search field: Again, see [VSCode's command palette](https://github.com/addaix/raglab/discussions/1#discussioncomment-8517049). This "smart-dynamic-filter" functionality should be able to used for our commands (on top of any menu or button-based access to commands), but can also be used for any tabular view and most nested views.


### Problem aspects: Dimensions, Characteristics, Categories

Note: Find a more comprehensive/general description of this here [Collection Explorers: Design Dimensions and Considerations](https://github.com/otosense/oui/discussions/14#:~:text=Collection%20Explorers%3A%20Design%20Dimensions%20and%20Considerations).

What different aspects should we have in mind when we think about store explorers? 
What language can we use to communicate about the different data sources/targets, GUI components, and the communication between them. 

For example, as far as the data is concerned:
* it could be flat or nested
* we could have a fix schema (structure+interpretation) or not
* the size (e.g. number of items) could be known, or not.

From a front-end point of view, not sure what the patterns are:
* implicit (e.g. fetch more data when end of scroll) or explicit (e.g. "next" button) navigation 
* lazy-loading (or not)

### Notes about how to organize this store explorer stuff. 


Then, we need to talk about SoC (layers) and expand on the specifications for each concern/layer.

Here's what I have in mind to start with. The layers are.
* **Data**: target/source data, as it is (physical aspects, formats, protocols, native interface, etc.). Is it local files, remote blobs, live streaming data, mongo, sql, etc.).
* **Python Operation Interface**. All operations on data should be done through a facade with maximum reuse between various use-case. Whenever possible (and not a ridiculous stretch) we chose our facade's base/inner_facing interface from python builtin interfaces. For example, from [the abstract methods of python collections](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes). This is where packages like [dol](https://github.com/i2mint/dol) and [creek](https://github.com/i2mint/creek) come in: To help you build these interfaces from any data particulars. 
* **REST interface**: The python facades are mapped to REST design patterns, automatically, via [py2http](https://github.com/i2mint/py2http) tools. As always, these interfaces should be mapped to JS functions ([http2js](https://github.com/i2mint/http2js)?) and those should be used in the JS code.
* **View patterns**: The REST web-service APIs are mapped to common, good UX, view patterns for store exploration. 

All of these concerns have core reusable tools as their final target, but in absence of an existing core tool, the prod team is asked to: 
(1) Do the best they can to follow good principles
(2) Communicate with the core team about this.

Core tools exist for the first link (data to python-facade). 

There's probably sufficient tools for the second link (python-facade to web-service), though it would be better to pack this logic into a ready-to-use tool. See [Mapping python collections builtin methods to REST API patterns](https://github.com/i2mint/py2http/discussions/9) about this.

The main missing piece here is the "REST interface (to data operations) to view patterns" link. 

Let's chose a few target "general view patterns" (like our current highly customizable smart tables and nested-tree explorers) and work on getting those with a consistent REST interface (though `http2js` interfaces) as the backend.

**Comment by thorwhalen** [comment]
## Integrating callbacks to rjsf forms


I have a function that generates a rjsf form specification for a given web service API endpoint. 
It spits out a (jsonizable) dict like this:

```
{'rjsf': {'schema': {'title': 'prompt',
   'type': 'object',
   'properties': {'thing': {'type': 'string'},
    'n': {'type': 'string', 'default': '30'},
    'min_length': {'type': 'string', 'default': '1'},
    'max_length': {'type': 'string', 'default': '15'}},
   'required': ['thing'],
   'description': 'The function composes a prompt and asks an LLM to respond to it.'},
  'uiSchema': {'ui:submitButtonOptions': {'submitText': 'Run'},
   'thing': {'ui:autofocus': True}},
  'liveValidate': False,
  'disabled': False,
  'readonly': False,
  'omitExtraData': False,
  'liveOmit': False,
  'noValidate': False,
  'noHtml5Validate': False,
  'focusOnFirstError': False,
  'showErrorList': 'top'}}
```

where should I specify what should happen when the user clicks the "Submit" button.
Namely, I'd like the form data to be sent as the pay load of a POST request of my targeted API endpoint.

### Answer

To integrate the functionality of submitting form data as the payload of a POST request to a targeted API endpoint within a React JSON Schema Form (rjsf), it's necessary to handle this process programmatically because the rjsf library primarily deals with the generation and validation of forms based on JSON Schema. The actual submission logic, especially when it involves interacting with an external API, needs to be implemented separately, typically in the form's `onSubmit` event handler.

Given the structure of your rjsf form specification, here's an outline of how you can extend it to include submission functionality:

1. **Form Component Definition**: Ensure your form component is properly set up to use the rjsf library, importing the necessary components and defining the schema and UI schema based on your provided specification.

2. **Submission Handler Implementation**: Implement an `onSubmit` function within your form component. This function will be triggered when the user clicks the "Submit" button. Inside this function, you will collect the form data and then use it to make a POST request to your targeted API endpoint.

3. **Integrating the Submission Handler**: Modify the rjsf form component in your React component to include the `onSubmit` prop, passing your submission handler function as its value.

Here's a simplified example to illustrate these steps:

### Step 1: Import rjsf and Define Your Component

```jsx
import React from 'react';
import Form from '@rjsf/core';

// Assuming yourFormSpec is the variable holding the form specification shown above
const yourFormSpec = {/* your form specification object */};

const MyFormComponent = () => {
  // Step 2: Implementation of the Submission Handler
  const handleSubmit = ({formData}, e) => {
    // Use fetch or any HTTP client to send formData to your API
    fetch('YOUR_API_ENDPOINT', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(formData),
    })
    .then(response => response.json())
    .then(data => {
      console.log('Success:', data);
      // Handle success, possibly informing the user or redirecting
    })
    .catch((error) => {
      console.error('Error:', error);
      // Handle error, possibly informing the user
    });
  };

  // Step 3: Integrating the Submission Handler with the rjsf Form
  return (
    <Form
      schema={yourFormSpec.rjsf.schema}
      uiSchema={yourFormSpec.rjsf.uiSchema}
      onSubmit={handleSubmit}
      // You can also pass other props here as necessary
    />
  );
};
```

### Important Notes:

- **YOUR_API_ENDPOINT**: Replace this with the actual URL of your API endpoint where you intend to submit the form data.
- **Error Handling**: The example includes basic error logging. You'll likely need to implement more user-friendly error handling based on your application's requirements.
- **Success Handling**: Similarly, consider how you want to handle successful submissions. This could involve displaying a success message, redirecting the user to another page, or resetting the form.
- **Security**: Ensure that your API endpoint is secured appropriately to handle incoming requests, and consider any necessary authentication mechanisms.

This approach allows you to leverage the rjsf library for form generation and validation, while also providing the flexibility to integrate with any backend service by customizing the submission logic according to your needs.

**Reply by thorwhalen** [reply]
Note that the above is just chatGPT's answer to my query, and should be taken with a grain of salt.

For example, we want to make sure we make our "function calling" component so that 
(1) the callbacks such as `handleSubmit` (and other callbacks that might be used in the form (such as to get lists of valid inputs for some fields) should be make easy to "inject" into the component
(2) callbacks that essentially wrap http requests should be made easy to produce as well

Perhaps looking at what this [FunctionCaller](https://github.com/otosense/ogui/blob/main/src/JsonForm/FunctionCaller.tsx) does could help?

**Comment by thorwhalen** [comment]
## Persistent Identifiers

_Some data management systems, like github issues and discussions for example, don't delete items when moved or renamed, but instead use links. For example, moving an issue will keep the original id (and url) of the object, but it will forward to the new location, and the same item can then be access via the new url, but also the old one. What is this method called?_

This method doesn't have a single universally recognized name, but there are a few terms that describe it well:

Persistent Identifiers (PIDs): This is a general term for identifiers that stay the same even when the location of the resource they point to changes. In this case, the original ID of the issue acts as a PID.

Content Addressable Storage (CAS): This is a storage system where data is identified by its content (like a hash) rather than its location. While not a direct fit, the concept of accessing data based on a unique identifier regardless of location aligns with this method.

Redirection with Code: HTTP redirects (301 - Moved Permanently) are often used to implement this behavior. These redirects point the user (or system) to the new location while keeping the original URL functional.

Canonicalization: This refers to the process of identifying the "main" version of a resource when there are duplicates. In this context, the original URL can be considered the canonical one, even though it redirects to the new location.

So, while there isn't one specific term, these terms all capture different aspects of the method you described: using persistent identifiers and redirects to maintain access to data even when it's moved or renamed.

**Comment by davidaleprog** [comment]
Je pose un use case intéressant que je remarque : 
Souvent quand j'explore un sujet, un mécanisme, une méthodologie ... Bref quelque chose avec des liens de cause à effet .

J'aime bien faire un schéma bloc en mode mind map.
Quand je demande à GPT ce genre de chose soit un me fait un truc en mode text genre A---> B ---->c 
soit il me fait une image ou le texte n'est même pas une langue connue.
Bref il n'y a pas d'entre 2. 

Il y a peut être quelque chose à faire pour générer un graph de connaissance à partir d'un texte... 
Toujours l'idée de Unstructured data à structured

**Reply by thorwhalen** [reply]
Ask it to give you a diagram using a diagraming language. For example graphviz or mermaid.

**Reply by thorwhalen** [reply]
You're entering the domain of what the grand POC velocity idea is. Let's talk more about it.

**Reply by thorwhalen** [reply]
It goes more or less like this:

```mermaid
flowchart
  ideas --> diagrams & specifications
  diagrams & specifications & ai--> poc_code
  poc_code & further_details & more_ai--> mvp_code
```

And note that I did that here, within github, with this code (see the raw text behind this post)

```
flowchart
  ideas --> diagrams & specifications
  diagrams & specifications & ai--> poc_code
  poc_code & further_details & more_ai--> mvp_code
```

Also know that there's some "plugins" to do this in chatGPT (render diagrams for you). 
But for sure, you can tell chatGPT "give me a diagram for this in the graphviz format (or the mermaid one, but I find graphviz more flexible if I want to extend...".
